Unbind an instance from the cluster.

jimwebber opened the pull request:
Current implementation merely checks that the database exists,
and that it isn't locked by an active Neo4j instance.

If all is good then it deletes the cluster-state directory reverting the
store to a Neo4j standalone/ha format (which can be used to
seed another core-edge cluster).

jimwebber committed:
Unbind an instance from the cluster.

Current implementation merely checks that the database exists,
and that it isn't locked by an active Neo4j instance.

If all is good then it deletes the cluster-state directory reverting the
store to a Neo4j standalone/ha format (which can be used to
seed another core-edge cluster).

srbaker commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,184 @@
+        try
+        {
+            Path pathToSpecificDatabase =
+                    homeDir.resolve( "data" ).resolve( "databases" ).resolve( databaseNameFrom( parsedArgs ) );
```
This will not work if the user has changed the `data` directory in configuratoin. It's probably better to load config, and use `GraphDatabaseSettings.database_path`. See `CheckConsistencyCommand` for example.

jimwebber replied:
Done.

srbaker commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,184 @@
+        String databaseName = parsedArgs.get( "database" );
+        if ( databaseName == null )
+        {
+            throw new IllegalArgumentException(
```
We use `IncorrectUsage` in `AdminCommand`s. `CheckConsistencyCommand` is another example of this pattern.

srbaker commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,184 @@
+        {
+            throw new IllegalArgumentException(
+                    "No database name specified. Usage: neo4j-admin " + "unbind --database-<name>" );
+        }
```
Usage information gets shown on `IncorrectUsage` errors, which is thrown above. So you can shrink this to: `return parsedArgs.interpretOption( "database", Converters.mandatory(), s -> s );`

jimwebber replied:
Okey.

srbaker commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,184 @@
+        }
+    }
+
+    private void confirmTargetDatabaseDirectoryExists( Path target )
```
You can probably use `Validators.CONTAINS_EXISTING_DATABASE` for this.

jimwebber replied:
Okey.

benbc commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,184 @@
+        Path lockFile = dbDir.resolve( StoreLocker.STORE_LOCK_FILENAME );
+        if ( exists( lockFile ) )
+        {
+            if ( Files.isWritable( lockFile ) )
```
You fail silently if the lockfile is unwritable (i.e. if the current user doesn't have the correct permissions to operate on the database). Presumably that's not what you want for this command. This case will happen, for example, with the Debian package if the operator forgets to `sudo -u neo4j`.

jimwebber replied:
The existing `DumpCommand` code is where I started. I've merged my additions into a common abstraction now (`StoreLockChecker`).

benbc replied:
Where is that?

jimwebber replied:
It's there now. I had to get on a plane prior to pushing.

benbc commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,184 @@
+                StoreLocker storeLocker = new StoreLocker( fsa );
+                try
+                {
+                    storeLocker.checkLock( dbDir.toFile() );
```
You don't unlock anywhere. I _think_ that's going to be okay with the current implementation because we're sure that after this command is done, the JVM will just exit. But since this class doesn't have a main method and is instead invoked by code from elsewhere, I don't think it's reasonable for the implementation to rely on an immediate exit. You may also run into problems with test ordering.

benbc reviewed (changes requested):
Could you not save yourself some effort by reusing the existing lock handling code from `DumpCommand`?

jimwebber committed:
Unbind an instance from the cluster.

Using the neo4j config to determine the database location.
Re-using existing validator code, and deleted proprietary version of same.

jimwebber committed:
Unbind an instance from the cluster.

Embed the non-writable directory in the CannotWriteException for
sane debugging.

benbc commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,180 @@
+    private void confirmTargetDirectoryIsWritable( Path pathToSpecificDatabase )
+            throws CommandFailed, CannotWriteException
+    {
+        new StoreLockChecker().withLock( pathToSpecificDatabase );
```
Do you have a reason for not unlocking the store once you're done?

maxsumrall commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,193 @@
+        @Override
+        public String description()
+        {
+            return "Removes cluster state data from the specified database making it suitable for use non-Core-Edge " +
```
"...suitable for use _in_ non-Core-Edge databases."

Forgot a word.

maxsumrall replied:
My suggestion that doesn't use the term `Core-Edge`:
"Removes cluster data from the specified database making it suitable to use as a single instance or to bind to a different cluster."

jimwebber replied:
Done.

maxsumrall commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,193 @@
+        catch ( StoreLockException e )
+        {
+            throw new CommandFailed(
+                    "Database is currently locked. Please check that no other Neo4j instance using it.", e );
```
"Please check that no other Neo4j instance _is_ using it."
Or, "Please check that it is not being used by another Neo4j instance."

benbc replied:
Isn't the most likely problem that _this_ database is running?

jimwebber replied:
@benbc do you have some concrete words that you'd prefer?

benbc replied:
The equivalent text for `neo4j-admin dump`, which I think is reasonable is "the database is in use -- stop Neo4j and try again". That identifies the most likely cause and gives clear instructions on what needs to be done to fix the problem.

maxsumrall commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,193 @@
+    {
+        if ( !exists( clusterStateDirectory ) )
+        {
+            throw new UnbindFailureException( "Database %s is not bound to any cluster", clusterStateDirectory );
```
Should we add some message saying that everything is fine? For example: "Unbinding not necessary."

benbc replied:
Does this surface as an error? I would expect this command to be idempotent -- and therefore exit 0 in this case, possible (but not definitely) with a notification that nothing was done.

jimwebber replied:
It's now idempotent. If there's no `cluster-state` directory then we'll skip the remaining steps.

maxsumrall commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,193 @@
+public class UnbindFromClusterCommand implements AdminCommand
+{
+
+    private StoreLockChecker storeLockChecker;
```
Can be final

jimwebber replied:
String no longer appears in code.

jimwebber committed:
Unbind an instance from the cluster.

Refining the error messages from PR feedback.

benbc commented on enterprise/core-edge/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -0,0 +1,181 @@
+            else
+            {
+                outsideWorld.stdOutLine(
+                        format( "No cluster state found in %s. No work perfomed.", pathToSpecificDatabase ) );
```
Informational output like this should go to stderr.

jimwebber replied:
Done.

jimwebber committed:
Unbind an instance from the cluster.

Change informational output from stdout to stderr.
